packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/service/storage_s3_service_impl.dart

```diff
@@ -0,0 +1,685 @@
+    });
+
+    try {
+      // The current capability of the `remove` API doesn't require parsing
+      // the [DeleteObjectOutput] returned by [S3Client.deleteObject].
+      return await s3client.deleteObject(request).result;
+    } on smithy.UnknownSmithyHttpException catch (error) {
+      // S3Client.deleteObject may return 403, for deleting a non-existing
+      // object, the API call returns a successful response
+      throw S3Exception.fromUnknownSmithyHttpException(error);
+    }
+  }
+
+  static String _getS3EndpointHost({required String region}) =>
+      endpoint_resolver.endpointResolver
+          .resolve(
+            endpoint_resolver.sdkId,
+            region,
+          )
+          .endpoint
+          .uri
+          .host;
+
+  /// Resolve a client object key to a "full" object key with proper prefix.
+  ///
+  /// This API is only used internally.
+  @internal
+  static Future<String> getResolvedPrefix({
+    required S3PrefixResolver prefixResolver,
+    required AWSLogger logger,
+    required StorageAccessLevel accessLevel,
+    String? identityId,
+  }) async {
+    try {
+      return await prefixResolver.resolvePrefix(
+        accessLevel: accessLevel,
+        identityId: identityId,
+      );
+    } on Exception catch (error, st) {
+      logger.error('Error happened while resolving prefix', error, st);
+      throw S3Exception(
+        'Error happened while resolving prefix.',
+        recoverySuggestion:
+            'If you are providing a custom prefix resolver, please review the underlying exception to determine the cause.',
+        underlyingException: error,
+      );
+    }
+  }
+
+  /// Creates and sends a [s3.HeadObjectRequest] to S3 service, and then
+  /// returns a [s3.HeadObjectOutput].
+  ///
+  /// This API is only used internally.
+  @internal
+  static Future<s3.HeadObjectOutput> headObject({
+    required s3.S3Client s3client,
+    required String bucket,
+    required String key,
+  }) async {
+    final request = s3.HeadObjectRequest.build((builder) {
+      builder
+        ..bucket = bucket
+        ..key = key;
+    });
+
+    try {
+      return await s3client.headObject(request).result;
+    } on smithy.UnknownSmithyHttpException catch (error) {
+      // S3Client.headObject may return 403 or 404 error
+      throw S3Exception.fromUnknownSmithyHttpException(error);
+    }
+  }
+
+  /// As background operation is not yet supported, in order to avoid incomplete
+  /// multipart uploads (caused by App process being killed) from occupying
+  /// storage space, [s3.AbortMultipartUploadRequest]s are sent for each
+  /// multipart upload id stored in the [transfer.TransferDatabase] to remove
+  /// uploaded parts upon [StorageS3Service] initiation.
+  @internal
+  Future<void> abortIncompleteMultipartUploads() async {
+    final records = await _transferDatabase
+        .getMultipartUploadRecordsCreatedBefore(_serviceStartingTime);
+
+    for (final record in records) {
```

Comment:
Can these be parallelized w/ `Future.wait`?

Reply:
Was thinking about this, this is a low priority task, thought executing 1 by 1 will give more resource to other http requests? (if upload/download already started). Thoughts?

Reply:
I think it's probably alright. If this is being awaited elsewhere, it may just be creating an unnecessary delay.

Reply:
👍 this is `unawaited` in S3 plugin configure.